### PR TITLE
refactor: implement MrVI in PyTorch

### PR DESCRIPTION
**Pull request by lordy5**
Currently finished first rough implementation in torch, but need to make a lot of fixes and debug still.

### CODE REVIEW

**Comment by ori-kron-wis** [comment]
In this (big) PR we also make the jax (and perhaps pyro) optional - that affects many models' dependency warning. 
Added a duplication of old mrVI Jax version for reference/transition period until deprecated, if any (+its tests). 
Added changelog, mrVi test will run from now on (needed to put their label)

**Comment by ori-kron-wis** [comment]
close also https://github.com/scverse/scvi-tools/issues/2349

File: src/scvi/external/mrvi_torch/_model.py [hunk 1]
```diff
@@ -0,0 +1,1679 @@
...
+    "plan_kwargs": {
+        "lr": 2e-3,
+        "n_epochs_kl_warmup": 20,
+        "eps": 1e-8,
+        "weight_decay": 1e-8,
+    },
+}
+
+
+class TorchMRVI(VAEMixin, UnsupervisedTrainingMixin, BaseModelClass):
```
[the remark this answers is not in the document]
Comment (ori-kron-wis) [reply src/scvi/external/mrvi_torch/_model.py:51]: Its a good question and we actually had this way before, but decided to change back to the current situation just for users backcompatibility between versions (i.e MRVI is still JAX and they will be able to load their previous models naturally, scripts will run and so on). Hopefully those warnings will be enough for this transition period.

File: src/scvi/external/mrvi_torch/_module.py [hunk 2]
```diff
@@ -0,0 +1,608 @@
...
+        u_stop = u if not self.stop_gradients else u.detach()
+        u_ = self.layer_norm(u_stop)
+
+        sample_embed = self.layer_norm_embed(self.embedding(sample_covariate))
+        # TODO: WHY NOISE WAS ADDED HERE?
+        # Add noise to sample embedding using proper random sampling instead of in-place init
+        noise = torch.normal(
+            mean=0.0, std=0.1, size=sample_embed.shape, device=sample_embed.device
+        )
+        sample_embed = sample_embed + noise
```
[the remark this answers is not in the document]
Comment (ori-kron-wis) [reply src/scvi/external/mrvi_torch/_module.py:247]: Seems to me to just remove this code block, as init.normal_(self.embedding.weight, std=0.1) exists after init of the embedding, like in jax, and like in other encoders/decoder. This is the only encoder the noise is added afterwards as well.

[the remark this answers is not in the document]
Comment (PierreBoyeau) [reply src/scvi/external/mrvi_torch/_module.py:247]: Sounds good!

File: src/scvi/external/mrvi_torch/_model.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (ori-kron-wis) [reply src/scvi/external/mrvi_torch/_model.py:51]: we decided to move on with a backend parameter ("jax" or "torch") to the MRVI class

File: src/scvi/dataloaders/_custom_dataloders.py [hunk 3]
```diff
@@ -287,6 +347,20 @@ def registry(self) -> dict:
                     },
                     "summary_stats": {"n_labels": self.n_labels},
                 },
+                "ind_x": {
```
[the remark this answers is not in the document]
Comment (ori-kron-wis) [reply src/scvi/dataloaders/_custom_dataloders.py:350]: yes, we use hard-coded registry to support the custom dataloaders. I agree we need to reconsider in a later PR (not only for mRVI but more generally)

File: src/scvi/external/mrvi_torch/_components.py [hunk 4]
```diff
@@ -0,0 +1,306 @@
...
+    from typing import Literal
+
+import torch
+from torch import nn
+from torch.distributions import Normal
+
+PYTORCH_DEFAULT_SCALE = 1 / 3
+
+
+class Dense(nn.Linear):
```
[the remark this answers is not in the document]
Comment (ori-kron-wis) [reply src/scvi/external/mrvi_torch/_components.py:16]: just a legacy from the jax implementation i suppose, but not really needed now
